tensorflow/core/kernels/adjust_hue_op_gpu.cu.cc

```diff
@@ -0,0 +1,147 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#if GOOGLE_CUDA
+
+#define EIGEN_USE_GPU
+
+#include "tensorflow/core/kernels/adjust_hsv_op_common.h"
+#include "tensorflow/core/util/cuda_kernel_helper.h"
+
+namespace tensorflow {
+
+typedef Eigen::GpuDevice GPUDevice;
+
+namespace internal {
+
+typedef struct RgbTuple {
+  float r;
+  float g;
+  float b;
+} RgbTuple;
+
+typedef struct HsvTuple {
+  float h;
+  float s;
+  float v;
+} HsvTuple;
+
+__device__ HsvTuple rgb2hsv_cuda(const float r, const float g, const float b) {
+  HsvTuple tuple;
+  const float M = fmaxf(r, fmaxf(g, b));
+  const float m = fminf(r, fminf(g, b));
+  const float chroma = M - m;
+  float h = 0.0f, s = 0.0f;
+  // hue
+  if (chroma > 0.0f) {
+    if (M == r) {
+      const float num = (g - b) / chroma;
+      const float sgn = num < 0.0f;
+      const float sign = powf(-1.0f, sgn);
+      h = (sgn * 6.0f + sign * fmodf(sign * num, 6.0f)) / 6.0f;
+    } else if (M == g) {
+      h = ((b - r) / chroma + 2.0) / 6.0f;
+    } else {
+      h = ((r - g) / chroma + 4.0) / 6.0f;
+    }
+  } else {
+    h = 0.0f;
+  }
+  // saturation
+  if (M > 0.0) {
+    s = chroma / M;
+  } else {
+    s = 0.0f;
+  }
+  tuple.h = h;
+  tuple.s = s;
+  tuple.v = M;
+  return tuple;
+}
+
+__device__ RgbTuple hsv2rgb_cuda(const float h, const float s, const float v) {
+  RgbTuple tuple;
+  const float new_h = h * 6.0f;
+  const float chroma = v * s;
+  const float x = chroma * (1.0 - fabsf(fmodf(new_h, 2.0f) - 1.0f));
+  const float new_m = v - chroma;
+  const bool between_0_and_1 = new_h >= 0.0 && new_h < 1;
+  const bool between_1_and_2 = new_h >= 1.0 && new_h < 2;
+  const bool between_2_and_3 = new_h >= 2 && new_h < 3;
+  const bool between_3_and_4 = new_h >= 3 && new_h < 4;
+  const bool between_4_and_5 = new_h >= 4 && new_h < 5;
+  const bool between_5_and_6 = new_h >= 5 && new_h < 6;
+  tuple.r = chroma * (between_0_and_1 || between_5_and_6) +
+    x * (between_1_and_2 || between_4_and_5) + new_m;
+  tuple.g = chroma * (between_1_and_2 || between_2_and_3) +
+    x * (between_0_and_1 || between_3_and_4) + new_m;
+  tuple.b =  chroma * (between_3_and_4 || between_4_and_5) +
+    x * (between_2_and_3 || between_5_and_6) + new_m;
+  return tuple;
+}
+
+__global__ void adjust_hue_nhwc(const int number_elements,
+                                const float * const input, float * const output, const float * const hue_delta) {
+  const float delta = hue_delta[0];
+  // multiply by 3 since we're dealing with contiguous RGB bytes for each pixel (NHWC)
+  const int idx = (blockDim.x * blockIdx.x + threadIdx.x) * 3;
+  // bounds check
+  if (idx > number_elements) {
+    return;
+  }
+  const HsvTuple hsv = rgb2hsv_cuda(input[idx], input[idx+1], input[idx+2]);
```

Comment:
@zheng-xq I actually haven't tested the shared memory case. It might be slightly better that way, let me know if you'd like me to try. If it's OK with you, I can do it while working on the next kernel I'm about to PR (saturation). Let me know what you think.